src/llvm_ts.cpp

```diff
@@ -287,10 +286,31 @@ static void submit_cpu(KernelType type, Func &&func, uint32_t width,
     jitc_task = new_task;
 }
 
-Task *LLVMThreadState::launch(Kernel kernel, uint32_t size,
-                              std::vector<void *> *kernel_params,
-                              uint32_t /* kernel_param_count */,
-                              const uint8_t* /* kernel_params_global */) {
+/// Temporary scratch space for scheduled tasks
+static std::vector<Task *> scheduled_tasks;
+
+void LLVMThreadState::barrier() {
+    if (scheduled_tasks.size() == 1) {
+        task_release(jitc_task);
+        jitc_task = scheduled_tasks[0];
+    } else {
+        jitc_assert(!scheduled_tasks.empty(),
+                    "jit_eval(): no tasks generated!");
+
+        // Insert a barrier task
+        Task *new_task = task_submit_dep(nullptr, scheduled_tasks.data(),
+                                         (uint32_t) scheduled_tasks.size());
+        task_release(jitc_task);
```

Comment:
Is it always the case that `jitc_task == scheduled_task.front()` (or maybe `.back()`)?
In this function, I feel like either `jitc_task` is being freed twice, or it's being dropped (= not included in the dependencies of `new_task`).
But I'm not familiar with the task system.

Reply:
How about: "All tasks in 'scheduled_tasks' have 'jitc_task' (if present) as parent. To create a barrier, we release 'jitc_task' and initialize it with a new dummy task that has all members of 'scheduled_tasks' as parents. This allows groups of kernel launches to run in parallel, while serializing subsequent groups."